### PR TITLE
Add support for async fit()

### PR DESCRIPTION
**Pull request by iquintero**
when calling fit(wait=False) it will return immediately. The training
job will carry on even if the process exits. by using attach() the
estimator can be retrieved by providing the training job name.

This fixes: https://github.com/aws/sagemaker-python-sdk/issues/4

### CODE REVIEW

**Comment by iquintero** [comment]
Note:

I have done a big refactor of how attach() works. This is the main change here, everything else is mostly tests.

File: src/sagemaker/amazon/amazon_estimator.py [hunk 1]
```diff
@@ -64,6 +64,20 @@ def data_location(self, data_location):
             data_location = data_location + '/'
         self._data_location = data_location
 
+    @classmethod
+    def _from_training_job(cls, init_params, hyperparameters, image, sagemaker_session):
```
Comment (mvsusp) [inline src/sagemaker/amazon/amazon_estimator.py]: Can you add Docstrings to this method?

File: src/sagemaker/amazon/amazon_estimator.py [hunk 2]
```diff
@@ -64,6 +64,20 @@ def data_location(self, data_location):
...
+
+        # The hyperparam names may not be the same as the class attribute that holds them,
+        # for instance: local_lloyd_init_method is called local_init_method. We need to map these
+        # and pass the correct name to the constructor.
+
+        for attribute, value in cls.__dict__.items():
+            if isinstance(value, hp):
```
Comment (mvsusp) [inline src/sagemaker/amazon/amazon_estimator.py]: Accessing __dict__ to find a class attribute shows that ```_from_training_job``` should not be a class method but an instance method.

File: src/sagemaker/estimator.py [hunk 3]
```diff
@@ -152,8 +152,46 @@ def fit(self, inputs, wait=True, logs=True, job_name=None):
...
+    @classmethod
+    def attach(cls, training_job_name, sagemaker_session=None):
+        """Attach to an existing training job.
+
+        Create an Estimator bound to an existing training job. After attaching, if
+        the training job has a Complete status, it can be ``deploy()`` ed to create
+        a SageMaker Endpoint and return a ``Predictor``.
+
+        If the training job is in progress, attach will block and display log messages
+        from the training job, until the training job completes.
```
Comment (mvsusp) [inline src/sagemaker/estimator.py]: Can you add some code examples here?

File: src/sagemaker/estimator.py [hunk 4]
```diff
@@ -152,8 +152,46 @@ def fit(self, inputs, wait=True, logs=True, job_name=None):
...
+
+        Args:
+            training_job_name (str): The name of the training job to attach to.
+            sagemaker_session (sagemaker.session.Session): Session object which manages interactions with
+                Amazon SageMaker APIs and any other AWS services needed. If not specified, the estimator creates one
+                using the default AWS configuration chain.
+            **kwargs: Additional kwargs passed to the :class:`~sagemaker.estimator.Estimator` constructor.
```
Comment (mvsusp) [inline src/sagemaker/estimator.py]: kwargs is not in the method signature

File: src/sagemaker/estimator.py [hunk 5]
```diff
@@ -152,8 +152,46 @@ def fit(self, inputs, wait=True, logs=True, job_name=None):
...
+
+        Returns:
+            sagemaker.estimator.Framework: ``Estimator`` with the attached training job.
+        """
+        sagemaker_session = sagemaker_session or Session()
+
+        if training_job_name is not None:
```
Comment (mvsusp) [inline src/sagemaker/estimator.py]:
```
if not training_job_name
```

File: src/sagemaker/estimator.py [hunk 6]
```diff
@@ -528,56 +566,24 @@ def hyperparameters(self):
...
+        framework_init_params['source_dir'] = json.loads(hyperparameters.get(DIR_PARAM_NAME))
+        framework_init_params['enable_cloudwatch_metrics'] = json.loads(
+            hyperparameters.get(CLOUDWATCH_METRICS_PARAM_NAME))
+        framework_init_params['container_log_level'] = json.loads(
+            hyperparameters.get(CONTAINER_LOG_LEVEL_PARAM_NAME))
 
         # drop json and remove other SageMaker specific additions
-        hyperparameters = {entry: json.loads(hp[entry]) for entry in hp}
-        framework_init_params['hyperparameters'] = hyperparameters
+        hp_map = {entry: json.loads(hyperparameters[entry]) for entry in hyperparameters}
```
Comment (mvsusp) [inline src/sagemaker/estimator.py]: hp is a map already. Maybe rename to deseriealized_hps or something in this lines?
My following up question is why is this method responsible to deserialize the hps?

File: tests/integ/test_factorization_machines.py [hunk 7]
```diff
@@ -53,3 +54,45 @@ def test_factorization_machines():
...
+
+        # training labels must be 'float32'
+        fm.fit(fm.record_set(train_set[0][:200], train_set[1][:200].astype('float32')), wait=False)
+        training_job_name = fm.latest_training_job.name
+
+        print("Detached from training job. Will re-attach in 20 seconds")
+        time.sleep(20)
+        print("attaching now...")
+
+    with timeout_and_delete_endpoint_by_name(endpoint_name, sagemaker_session, minutes=35):
```
Comment (mvsusp) [inline tests/integ/test_factorization_machines.py]: 35 minutes is too long. Can we make a test under 15 minutes?

File: tests/integ/test_kmeans.py [hunk 8]
```diff
@@ -60,3 +61,49 @@ def test_kmeans():
...
+        kmeans.center_factor = 1
+
+        kmeans.fit(kmeans.record_set(train_set[0][:100]), wait=False)
+        training_job_name = kmeans.latest_training_job.name
+
+        print("Detached from training job. Will re-attach in 20 seconds")
+        time.sleep(20)
+        print("attaching now...")
+
+    with timeout_and_delete_endpoint_by_name(endpoint_name, sagemaker_session, minutes=35):
```
Comment (mvsusp) [inline tests/integ/test_kmeans.py]: Too long

File: tests/integ/test_factorization_machines.py (hunk 7 above)
Comment (iquintero) [reply tests/integ/test_factorization_machines.py]: The test itself is taking the same time as the synchronous fit one, the difference is this timeout is accounting for fit() + deploy(). If you notice the timeout is consistent with the synchronous tests that we have.

The difference is the call to

with timeout_and_delete_endpoint_by_name() is spending a lot of time on attach()  this time is usually spent in fit() in the synchronous tests. In this case the fit() returns right away so it doesn't account for much of the runtime. So the test takes the same amount of time to complete.

File: src/sagemaker/amazon/amazon_estimator.py (hunk 1 above)
Comment (iquintero) [reply src/sagemaker/amazon/amazon_estimator.py]: will do

File: src/sagemaker/estimator.py (hunk 4 above)
Comment (iquintero) [reply src/sagemaker/estimator.py]: will get rid of it. It used to be and I forgot to get rid of it here.

File: src/sagemaker/estimator.py (hunk 5 above)
Comment (iquintero) [reply src/sagemaker/estimator.py]: will do

File: src/sagemaker/estimator.py (hunk 3 above)
Comment (iquintero) [reply src/sagemaker/estimator.py]: will add an example.

File: src/sagemaker/estimator.py (hunk 6 above)
Comment (iquintero) [reply src/sagemaker/estimator.py]: I will rename it :D
- this was already done by attach(). I don't think its worth to move this deserialization to its own method.

File: src/sagemaker/amazon/amazon_estimator.py (hunk 2 above)
Comment (owen-t) [reply src/sagemaker/amazon/amazon_estimator.py]: I think this is fine here, hyperparameters are implemented as descriptor objects.

File: src/sagemaker/estimator.py [hunk 9]
```diff
@@ -152,8 +152,46 @@ def fit(self, inputs, wait=True, logs=True, job_name=None):
         self.latest_training_job = _TrainingJob.start_new(self, inputs)
         if wait:
             self.latest_training_job.wait(logs=logs)
+
+    @classmethod
+    def _from_training_job(cls, init_params, hyperparameters, image, sagemaker_session):
```
Comment (owen-t) [inline src/sagemaker/estimator.py]: Add documentation here, it's important. You're introducing a protocol that subclasses need to follow, so this should be documented.

File: src/sagemaker/estimator.py [hunk 10]
```diff
@@ -152,8 +152,46 @@ def fit(self, inputs, wait=True, logs=True, job_name=None):
...
+        """
+        sagemaker_session = sagemaker_session or Session()
+
+        if training_job_name is not None:
+            job_details = sagemaker_session.sagemaker_client.describe_training_job(TrainingJobName=training_job_name)
+            init_params, hp, image = cls._prepare_estimator_params_from_job_description(job_details)
+
         else:
-            raise NotImplemented('Asynchronous fit not available')
+            raise ValueError('must specify training_job name')
```
Comment (owen-t) [inline src/sagemaker/estimator.py]: IMO, this isn't necessary, just let the underlying call fail.

**Pull request merged**